core/src/main/java/dk/alexandra/fresco/framework/builder/numeric/DefaultPreprocessedValues.java

```diff
@@ -22,29 +23,19 @@ public DefaultPreprocessedValues(ProtocolBuilderNumeric builder) {
     return builder.seq(b -> {
       DRes<SInt> r = b.numeric().randomElement();
       DRes<SInt> inverse = b.advancedNumeric().invert(r);
-      ArrayList<DRes<SInt>> list = new ArrayList<>();
-      list.add(inverse);
-      list.add(r);
-      return () -> new IterationState(2, list);
-    }).whileLoop(
-        (state) -> state.round <= pipeLength + 1,
-        (seq, state) -> {
-          DRes<SInt> r = state.value.get(1);
-          DRes<SInt> last = state.value.get(state.value.size() - 1);
-          List<DRes<SInt>> values = new ArrayList<>(state.value);
-          values.add(seq.numeric().mult(last, r));
-          return () -> new IterationState(state.round + 1, values);
-        }).seq((seq, state) -> () -> state.value);
-  }
-
-  private static final class IterationState {
-
-    private final int round;
-    private final List<DRes<SInt>> value;
-
-    private IterationState(int round, List<DRes<SInt>> value) {
-      this.round = round;
-      this.value = value;
-    }
+      ArrayList<DRes<SInt>> values = new ArrayList<>(pipeLength + 2);
+      values.add(inverse);
+      values.add(r);
+      return () -> values;
+    }).whileLoop((values) -> values.size() <= pipeLength + 1, (seq, values) -> {
```

Comment:
And now the small print
Could you write < rather than <= and then drop the +1?

Reply:
No I could not. Because that would not be equivalent to what is there now. I could go with `... values.size() < pipeLength + 2 ... ` and save a character.